transportation-data-publishing/bcycle/bcycle_trip_pub.py

```diff
@@ -99,78 +66,66 @@ def handle_data(data):
-    dt_current_formatted = dt_current.format("MM-YYYY")
-    up_to_date = False
-    results = None
 
-    while not up_to_date:
+def main():
+    today = arrow.get(datetime.today())
+    # bcycle dropbox data is only ever available for the previous month
+    max_socrata_dt = today.shift(months=-1)
+    # we want to test against the first day of the last month, so construct that
+    results = 0
 
-        socrata_dt = max_date_socrata(resource_id)
+    while True:
 
-        socrata_dt_formatted = arrow.get(socrata_dt).format("MM-YYYY")
+        current_dt_socrata = arrow.get(get_newest_socata_record(resource_id))
 
-        if dt_current_formatted == socrata_dt_formatted:
+        if current_dt_socrata >= max_socrata_dt:
+            break
 
-            up_to_date = True
+        # try to download the file for month for which we do not have data
+        next_file_date = current_dt_socrata.shift(months=+1)
+        dropbox_file_dt = next_file_date.format("MMYYYY")
 
-            results = 0
-
-        else:
-            #  socrata data is at least one month old
-            dropbox_month = arrow.get(socrata_dt).replace(months=1).format("MM")
-            dropbox_year = arrow.get(socrata_dt).replace(months=1).format("YYYY")
-
-            current_file = "TripReport-{}{}.csv".format(dropbox_month, dropbox_year)
-            root = "austinbcycletripdata"  # note the lowercase-ness
-            path = "/{}/{}/{}".format(root, dropbox_year, current_file)
-            date_fields = ["checkout_date"]
-            try:
-                try:
-                    data = get_data(path, DROPBOX_BCYCLE_TOKEN)
-                    results = len(data)
-
-                except TypeError:
-                    results = 0
-
-            except dropbox.exceptions.ApiError as e:
-
-                if "LookupError" in str(e):
-                    # end loop when no file can be found
-                    up_to_date = True
-                    break
-                else:
-                    raise e
-
-            data = handle_data(data)
-
-            socratautil.Soda(
-                auth=SOCRATA_CREDENTIALS,
-                records=data,
-                resource=resource_id,
-                location_field=None,
-                source="bcycle"
-            )
+        current_file = "TripReport-{}.csv".format(dropbox_file_dt)
+        root = "austinbcycletripdata"  # note the lowercase-ness
+        path = "/{}/{}/{}".format(root, next_file_date.year, current_file)
 
+        try:
+            data = get_data(path, DROPBOX_BCYCLE_TOKEN)
             results = len(data)
+        except TypeError:
+            results = 0
+        except dropbox.exceptions.ApiError as e:
+
+            if "LookupError" in str(e):
+                # end loop when no file can be found
+                break
+            else:
+                raise e
+
+        data = handle_data(data)
+        data = format_checkout_date(data, next_file_date.year)
```

Comment:
Is the date data always provided? If not, shouldn't errors be handled in a try-catch statement here or within the function? Also, what would happen (for example) if the date is empty, would `format_checkout_date ` and socratautil be ok having an empty date?

Reply:
@sergiogcx yes, appreciate that concern. these CSV's are created by B-Cycle staff and uploaded to dropbox, and the entire ETL is dependent on that CSV being formatted exactly according to the format we've agreed upon. so this whole process is very brittle. 

to answer your question: if the date is empty, the script should and will fail. every record must have a checkout date.